pyccel/ast/numpyext.py

```diff
@@ -722,24 +726,71 @@ def is_elemental(self):
 #==============================================================================
 class NumpyWhere(PyccelInternalFunction):
     """ Represents a call to  numpy.where """
-    __slots__ = ()
+
+    __slots__ = ('_condition', '_value_true', '_value_false', '_dtype',
+                 '_rank', '_shape', '_order', '_precision')
+    _attribute_nodes = ('_condition','_value_true','_value_false')
     name = 'where'
 
-    def __init__(self, mask):
-        super().__init__(mask)
+    def __new__(cls, condition, x = None, y = None):
+        if x is None and y is None:
+            return NumpyNonZero(condition)
+        else:
+            return super().__new__(cls)
+
+    def __init__(self, condition, x = None, y = None):
+        self._condition = condition
+        self._value_true = x
+        self._value_false = y
 
+        args      = (x, y)
+        integers  = [e for e in args if e.dtype is NativeInteger() or e.dtype is NativeBool()]
+        floats    = [e for e in args if e.dtype is NativeFloat()]
+        complexs  = [e for e in args if e.dtype is NativeComplex()]
+
+        if complexs:
+            self._dtype     = NativeComplex()
+            self._precision = max(e.precision for e in complexs)
```

Comment:
Are we sure about how we calculate the precision?

We could have a complex `x` with low precision, and a float `y` with very high precision. With our current choice we lose precision. Here is an example:

```python
In [1]: import numpy as np

In [2]: a = np.complex64(3)

In [3]: a
Out[3]: (3+0j)

In [4]: b = np.float128(4)

In [5]: b
Out[5]: 4.0

In [6]: c = a + b

In [7]: c.dtype
Out[7]: dtype('complex256')

In [8]: c
Out[8]: (7+0j)
```

Reply:
You are right. This is also true in all of our operators. I have fixed it here : [60a38b3](https://github.com/pyccel/pyccel/pull/858/commits/60a38b357f32e8763dacdf32c32b9e1f10ee4577) but I will open a separate PR for the rest as there is already too much in this PR